### PR TITLE
fix(remix-dev): flat route fixes and enhancements

### PR DESCRIPTION
**Pull request by mcansh**
Signed-off-by: Logan McAnsh <logan@mcan.sh>

Closes: #5170
Closes: #5261
Closes: #5303
Closes: #5322
Closes: #5324
Closes: #5327

- [ ] Docs
- [x] Tests

Testing Strategy:

updated current tests

### CODE REVIEW

File: packages/remix-dev/config/flat-routes.ts [hunk 1]
```diff
@@ -124,6 +131,18 @@ export function getRouteSegments(routeId: string) {
...
+  /**
+   * @see https://github.com/remix-run/remix/pull/5160#issuecomment-1402157424
+   */
+  if (hasFolder && (routeId.endsWith("/index") || routeId.endsWith("/route"))) {
+    let last = routeId.lastIndexOf(path.posix.sep);
+    if (last >= 0) {
+      routeId = routeId.substring(0, last);
+    }
+  }
+
```
Comment (mcansh) [inline packages/remix-dev/config/flat-routes.ts:145]: check if we have a routeId with a folders, if we do, drop `index` and `route` segments

File: packages/remix-dev/config/flat-routes.ts [hunk 2]
```diff
@@ -33,6 +33,10 @@ interface RouteInfo extends ConfigRoute {
   segments: string[];
 }
 
+// prettier-ignore
+let routeRegex = /(([+][/\\][^/\\:?*]+)|[/\\](index|route)|(_[^/\\:?*]+))\.(tsx?|jsx?|mdx?)$$/;
```
Comment (mcansh) [inline packages/remix-dev/config/flat-routes.ts:37]: i understand... some of this regex...

https://github.com/remix-run/remix/pull/5034#discussion_r1065930461

**Comment by silvenon** [comment]
Fixed my issues perfectly! ❤️

**Comment by MajorTom327** [comment]
Same for me. Perfect !
Good job team ! 🎉

**Pull request merged**